crates/bevy_ui/src/render/gradient.rs

```diff
@@ -0,0 +1,984 @@
+                            i += 1;
+                        }
+
+                        extracted_gradients.items.push(ExtractedGradient {
+                            render_entity: commands.spawn(TemporaryRenderEntity).id(),
+                            stack_index: uinode.stack_index,
+                            transform: transform.compute_matrix(),
+                            stops_range: range_start..extracted_color_stops.0.len(),
+                            rect: Rect {
+                                min: Vec2::ZERO,
+                                max: uinode.size,
+                            },
+                            clip: clip.map(|clip| clip.clip),
+                            extracted_camera_entity,
+                            main_entity: entity.into(),
+                            node_type,
+                            border_radius: uinode.border_radius,
+                            border: uinode.border,
+                            resolved_gradient: ResolvedGradient::Radial { center: c, size },
+                        });
+                    }
+                    Gradient::Conic {
+                        position: center,
+                        stops,
+                    } => {
+                        let g_start = center.resolve(
+                            target.scale_factor,
+                            uinode.size,
+                            target.physical_size.as_vec2(),
+                        );
+                        let range_start = extracted_color_stops.0.len();
+
+                        // sort the explicit stops
+                        sorted_stops.extend(stops.iter().filter_map(|stop| {
+                            stop.angle
+                                .map(|angle| (stop.color.to_linear(), angle, stop.hint))
+                        }));
+                        sorted_stops.sort_by_key(|(_, angle, _)| FloatOrd(*angle));
+                        let mut sorted_stops_drain = sorted_stops.drain(..);
+
+                        // fill the extracted stops buffer
+                        extracted_color_stops.0.extend(stops.iter().map(|stop| {
+                            if stop.angle.is_none() {
+                                (stop.color.to_linear(), f32::NAN, stop.hint)
+                            } else {
+                                sorted_stops_drain.next().unwrap()
+                            }
+                        }));
+
+                        // interpolation
+                        let stops_out = &mut extracted_color_stops.0[range_start..];
+
+                        if stops_out[0].1.is_nan() {
+                            stops_out[0].1 = 0.;
+                        }
+                        if stops_out.last().unwrap().1.is_nan() {
+                            stops_out.last_mut().unwrap().1 = TAU;
+                        }
+
+                        let mut i = 1;
+
+                        while i < stops_out.len() - 1 {
```

Comment:
is this loop the same as the other two above? maybe extract into a function and share or highlight the differences

Reply:
Yep, initial implementation they did have differences but I simplified things down once I understood what I was doing and they are the same now more or less and should be extracted.